### DISCUSSION TITLE
When to set up the ports after performing a hard reset

### DISCUSSION THREAD

**Discussion by DoogieHowitzerWMD** [GerryDazoo/Slinger#248] [Q&A] [answered]
When in the process of setting up a new slingbox from factory default do I need to open the ports I want it to use in my router?

I managed to get one of my slingboxes mapped to the desired port after a hard reset, but a second one did not take to the desired port (interfering with the port mapped on a 3rd slingbox). I'm not sure what I've done differently.

When I've tried setting the ports and limiting outbound connections on my router at the very beginning (before establishing an internet connection and performing the bulk of the download) it prevents the slingbox from doing so. Beyond that, I'm not sure why one time it worked and one time it didn't.

We're pushing that deadline here, so any advice would be much appreciated.

**Edit** I have UPnP disabled, so that should not be the reason.

**Comment by GerryDazoo** [comment, marked as answer]
Not sure what you mean by "interfering with" there is no reason why two slingboxes can't use the same port number.

**Reply by DoogieHowitzerWMD** [reply]
> Not sure what you mean by "interfering with" there is no reason why two slingboxes can't use the same port number.

They can? I was unaware of that.

I was having an issue specifically with the two that shared the same port number, where I if I connected to one, disconnected from it, and tried to connect to the other one (all with Slings server, not the self-hosted one you've made for us), it wouldn't connect until I performed a soft reset. I figured it had to do with the shared port.

**Reply by sp222222** [reply]
sling servers won't work like that, but slinger will take same port

**Reply by DoogieHowitzerWMD** [reply]
> sling servers won't work like that, but slinger will take same port

Yeah, I just tried it and it seems to work that way. I am running the server in the same location as the slingboxes, and if I wasn't that might not be the case, as my router isnt allowing me to forward to different ip addresses to the same port.

The issue I seem to be having now is with my slingboxes that are connected via MOCA LAN. Those are on wired connections that make their way back to the main router through the coax, and those conistently seem to start a stream and then the stream will drop within 2 to 5 seconds. The two slingboxes I have wired directly to my main router have very stable streams.

I'm going to try different media players and maybe put one on Wi-Fi (from the main router, not the access point its wired to that also has Wi-Fi) and see if that makes a difference.

**EDIT:** seems to be dropping in a second with everything now... to the logs I go

**Reply by GerryDazoo** [reply]
Doogie: Yes that's right, you can't forward the same external port to two different internal IP addresses. 
So use a different external ports to forward to the slingboxes.
ie.  520101 -> slingbox1:5201
      520102 -> slingbox2:5201
But I guess my next question is if you're running the server locally with the slingboxes why are you messing with port forwarding on your router for the Slingboxes?

The MOCA stuff "should" work if you're not having configuration/hardware issues. I ran for many months using MOCA to get to my slingboxes in my media room from my main MOCA capable router in a different room upstairs.

**Reply by DoogieHowitzerWMD** [reply]
> Doogie: Yes that's right, you can't forward the same external port to two different internal IP addresses. So use a different external ports to forward to the slingboxes. ie. 520101 -> slingbox1:5201 520102 -> slingbox2:5201 But I guess my next question is if you're running the server locally with the slingboxes why are you messing with port forwarding on your router for the Slingboxes?

All of my previous testing had been with the server operating in a different location, so I was lost in the sauce with being focused on identifying the right port number for a specific slingbox.

I had been using the basic way selecting the device and the external port to forward, and I had assumed that this method was creating a 1 to 1 mapping with the slingbox's internal port and the external port on the router. For all I know though, this is just a simple mechanic for steering some unspecified internal port of that slingbox to the router's external port that was specified. I seemed to have been able to steer the first new slingbox I set up to use a specific external port (5205), and I was then able to use that same internal port number for that slingbox with the Slinger server. When I saw the second one had a port forwarding rule for the same external port number as another slingbox, I just anticipated problems.

Thinking about it now though, I'm not sure if any of that even mattered. Did I need to "identify" a specific internal port for the slingbox, or does the Slinger server work with any sort of internal port for a slingbox (in that typical 5000-ish range) that I specify in the config file?


> The MOCA stuff "should" work if you're not having configuration/hardware issues. I ran for many months using MOCA to get to my slingboxes in my media room from my main MOCA capable router in a different room upstairs.

I have no problems pinging the slingboxes or the Raspberry Pi running the server. But unless the logs are able to make a determination on the issue, I may have to look into identifying and troubleshooting some potentially deeper network problems that go beyond whether something can be pinged - whether that's network configuration, hardware, the use of MOCA, etc. That's the next thing I'll have to work out now that I have everything with the slingboxes themselves squared away before the deadline.

But, while dropping in 2 to 5 seconds is obviously undesirable, I'm not sure what length of time I should reasonably expect a stream to run for uninterrupted on a device. Would it be 5 to 6 hours before some sort of hiccup would end a stream, or would a device be able to keep viewing the same continuous stream almost indefinitely? 

Thanks again for everything and good luck with what you're likely dealing with now as the sun sets on Sling's servers.

**Reply by GerryDazoo** [reply]
The Slinger software will take any old port. So even if you have two slingboxes behind your router you can always get to them by defining two different port forwarding rules with different external port numbers and using those in your config.ini.
When I was first testing this I kept a stream up for three days (on my  LAN) just to make sure....
A little "hiccup" should be handled OK by the TCP protocol. So yes, I would think it should run forever.

**Reply by DoogieHowitzerWMD** [reply]
> The Slinger software will take any old port. So even if you have two slingboxes behind your router you can always get to them by defining two different port forwarding rules with different external port numbers and using those in your config.ini. When I was first testing this I kept a stream up for three days (on my LAN) just to make sure.... A little "hiccup" should be handled OK by the TCP protocol. So yes, I would think it should run forever.

If it could feasibly run indefinitely, then I likely have some things to fix on my network. I just set up a couple dedicated raspberry pi's at my parent's house, plugged into the tv's they have, with the goal of providing a mostly constant stream. Whenever there is a hiccup, I can reset it for them at anytime by SSH tunnelling into the Pi and restarting a vlc stream from the command line. I also have to use a program called "screen" to keep the process going after dropping the SSH connection.

That works as intended, but the stability of the stream is a bit inconsistent. I seem to be able to get at least 3 hours solidly (enough to watch a ballgame) on the Pi's, but sometimes it takes a couple tries to initially start the stream (that part could be just a VLC thing). Still, it's much better results than with my phone. Usually, VLC will give me just a peak at the screen before it drops, but occassionally, I will get a good 5 to 10 minutes out of it. MXPlayer seems like it connects (I see a stream titled "slingbox"), but the video and audio never show up.

For what it's worth, I am running with something like V 3.06 using the older method with separate config files for each box. If the consolidated config file provides a smoother experience, I might switch to that and grab the latest software package here, but I need to image the current SD card first before I make changes.

**Reply by GerryDazoo** [reply]
FOur things:
1) Please upgrade to the latest software. It's backwards compatible and has numerous bug fixes.
2) Whether you use separate config.ini's or the unified version won't  give you a smoother experience it just makes it easier to use because you only need one server port configured.
3) You should have mentioned you were running on a Pi. :-). From your issues can you confirm you using Solo's or Pro's? The computing overhead is much higher for these boxes because the code needs to peek into every video packet and decrypt some data if needed. You just might be running out of steam on the Pis. Run a "top" to see what's going on. If python CPU usage is pegged at 100% that would be bad.
4) Are you trying to do all this streaming over WiFi? A wired Ethernet connection should be more reliable.

**Reply by DoogieHowitzerWMD** [reply]
> FOur things:
> 
> 1. Please upgrade to the latest software. It's backwards compatible and has numerous bug fixes.
> 2. Whether you use separate config.ini's or the unified version won't  give you a smoother experience it just makes it easier to use because you only need one server port configured.
> 3. You should have mentioned you were running on a Pi. :-). From your issues can you confirm you using Solo's or Pro's? The computing overhead is much higher for these boxes because the code needs to peek into every video packet and decrypt some data if needed. You just might be running out of steam on the Pis. Run a "top" to see what's going on. If python CPU usage is pegged at 100% that would be bad.
> 4. Are you trying to do all this streaming over WiFi? A wired Ethernet connection should be more reliable.

Thank you for giving me some insight with this:

1. I will do that as soon as I can.
2. Ok, I'll keep things as is for my config files and port settings after I download the updated repository.
3. The Pi running the server is an old First gen model B+, and the one's dedicated for the streaming output are Pi 4's with 4GB RAM. I am not sure what "Solo's" or "Pro's" is referring to. **EDIT:** The slingboxes... duh. I am running four 500's.

From running "top" on one of the streaming Pi's (after restarting the VLC stream to make sure it hasn't dropped the stream), I see like 3 or 4 VLC processes, but the total cpu usage from them is about 20-25%, which seems like that should be ok.

Running top on the Pi that is running the server, I see a process with very high CPU usage (99%), then top seems to refresh the screen and I see the highest process only running at about 3%. When I press 'q' to exit top, I see the last two "refreshed" outputs from top, and the upper one has that process with 99% usage.

4. Right now for the streaming Pi's, one is wired and one is using WiFi. My parents have xfinity so I need to figure out a method of getting a wired connection to that spot. All of the rooms have Coax, but Xfinity does not use a MOCA setup like Verizon, so that is something I need to look into. The Pi that's connected wirelessly is the one that is getting use right now, as the Pi that has the wired LAN has display issues with the tv it's hooked up to (I need to find the right combination of those HDMI options in /boot/config.txt)

At the location of the slingboxes and the Pi running the server, everything is on a wired connection. I have FIOS, so some are directly connected to the main router, which has the ethernet WAN going to the ONT. Then, one slingbox is connected to an Ethernet adapter that uses MOCA over the coax, and another one is connected to a second, older FIOS router that is configured as an access point, also using MOCA over the coax.
